segmentstore/server/host/src/test/java/io/pravega/segmentstore/server/host/S3ClientWrapper.java

```diff
@@ -0,0 +1,279 @@
+import java.util.concurrent.ConcurrentMap;
+import lombok.SneakyThrows;
+import lombok.Synchronized;
+
+/**
+ * Client wrapper for S3JerseyClient. It uses local filesystem to implement extended S3 JAVA client APIs.
+ */
+class S3ClientWrapper extends S3JerseyClient {
+    private final AclMap aclMap;
+    private final String baseDir;
+    private final ConcurrentMap<String, ConcurrentMap<Integer, CopyPartRequest>> multipartUploads = new ConcurrentHashMap<>();
+
+    public S3ClientWrapper(S3Config s3Config, AclMap aclMap, String baseDir) {
+        super(s3Config);
+        this.aclMap = aclMap;
+        this.baseDir = baseDir;
+    }
+
+    @Override
+    public DeleteObjectsResult deleteObjects(DeleteObjectsRequest request) {
+        return new DeleteObjectsResult();
+    }
+
+    @Synchronized
+    @Override
+    public PutObjectResult putObject(PutObjectRequest request) {
+
+        if (request.getObjectMetadata() != null) {
+            request.setObjectMetadata(null);
+        }
+        try {
+            Path path = Paths.get(this.baseDir, request.getBucketName(), request.getKey());
+            Files.createDirectories(path.getParent());
+            Files.createFile(path);
+        } catch (IOException e) {
+            throw new S3Exception(e.getMessage(), 0);
+        }
+        PutObjectResult retVal = new PutObjectResult();
+        if (request.getAcl() != null) {
+            long size = 0;
+            if (request.getRange() != null) {
+                size = request.getRange().getLast() + 1;
+            }
+            aclMap.addNewObject(request.getKey(), new AclSize(request.getAcl(), size));
+        }
+        return retVal;
+    }
+
+    @Synchronized
+    @Override
+    @SneakyThrows(BadOffsetException.class)
+    public void putObject(String bucketName, String key, Range range, Object content) {
+
+        Path path = Paths.get(this.baseDir, bucketName, key);
+        try (FileChannel channel = FileChannel.open(path, StandardOpenOption.WRITE)) {
+
+            long startOffset = range.getFirst();
+            long length = range.getLast() + 1 - range.getFirst();
+            do {
+                long bytesTransferred = channel.transferFrom(Channels.newChannel((InputStream) content),
+                        range.getFirst(), range.getLast() + 1 - range.getFirst());
+                length -= bytesTransferred;
+                startOffset += bytesTransferred;
+            } while (length > 0);
+
+            AclSize aclKey = aclMap.getAclSizeForObject(key);
+            aclMap.updateObject(key, aclKey.withSize(range.getLast() + 1));
+        } catch (IOException e) {
+            throw new BadOffsetException(key, 0, 0);
+        }
+    }
+
+    @Synchronized
+    @Override
+    public void setObjectAcl(String bucketName, String key, AccessControlList acl) {
+        AclSize retVal = aclMap.getAclSizeForObject(key);
+        if (retVal == null) {
+            throw new S3Exception("NoObject", 404, "NoSuchKey", key);
```

Comment:
Since we have access to Apache Commons, perhaps this would help:
`org.apache.commons.httpclient.HttpStatus`